### PR TITLE
Generic asset chain with generic-asset-token-dealer

### LINKED ISSUE #1: Switch to generic asset pallet
- switching  out the native currency to generic asset pallet, keep the code in the repo as opposed to a dependency so to align with how the dex-parachain works  for development. 
- must have correct chain-spec
- test with JSON-RPC calls

### LINKED ISSUE #2: Transfer generic assets with token dealer
Generic token dealer should be able to transfer tokens between this parachain and relay chain. 

To address conversion rates between the parachain assets themselves and with relay chain asset, on transfers, the pallet asks for the rates defined in the runtime.   

Tests should only include parachain balance updates

### PR DESCRIPTION
**Pull request by whalelephant**
- closes #1 
- closes #2 

Other details:
- introduced test to token-dealer which manages generic-asset transfers
- include upward_message within token_dealer as opposed to using `cumulus_upward_message` crate as it is removed in v2
- Dedicated events for each transaction to/from relay, to/from para

### CODE REVIEW

**Review by iorveth — CHANGES_REQUESTED** [review]
Good work on creating xcmp messages for token transfers between parachains!
Here are some questions, that are reasonable to discuss.
1. I do not understand the purpose of having this rates
```
pub SpendingToRelayRate get(fn spending_to_relay_rate) config(): BalanceOf<T>;
  pub GenericToSpendingRate get(fn generic_to_spending_rate) config(): BalanceOf<T>;
```
- We do we need different currencies for transfers and rate between them? 
- How will we handle the case, when rate changes?

1. I don't sure, we can reuse upward/ downward  messages for token transfers.
`let msg = <T::UpwardMessage>::transfer(dest.clone(), amount.clone());`
You only have remark field in downward message, but not in upward, so seems we can`t reuse it. 
1. Also, i think, that we should switch from `generic-assets` module on generic parachain and use `Currency` trait, which respective provider can be balances for paying transaction fees.
It makes implementation more generic.
Also i think, that  basic assets  module for tokens on generic-parachain is much more convenient now, because we do not need mint/burn and other complicated functionality, we had with `generic-asset` pallet, which is already removed from the frame.

**Comment by whalelephant** [comment]
> I do not understand the purpose of having this rates
pub SpendingToRelayRate get(fn spending_to_relay_rate) config(): BalanceOf<T>;
  pub GenericToSpendingRate get(fn generic_to_spending_rate) config(): BalanceOf<T>;
We do we need different currencies for transfers and rate between them?
How will we handle the case, when rate changes?

This is actually a hack right now -  it is not really an exchange rate as such. 
The upward/downward message of `Balance` is from `polkadot_core_primitives` and the units seem to be different on the parachain when using polkadotJS for testing. i.e. sending 1 mirco unit from relay chain requires 1 mili unit transfer in the parachain.  If the rate changes, this is a change to the runtime in the parachain which will need an upgrade.
*Suggestion: Lets see what we decide on the last comment and we may or may not need to address this*

> don't sure, we can reuse upward/ downward messages for token transfers.
let msg = <T::UpwardMessage>::transfer(dest.clone(), amount.clone());
You only have remark field in downward message, but not in upward, so seems we can`t reuse it.

Yes we cannot, Upward message type is for the relay chain, where there is only the DOT (or other network tokens), Hence we have the xcmp messages for `TokenTransfer`, `TransferAssets`

> Also, i think, that we should switch from generic-assets module on generic parachain and use Currency trait, which respective provider can be balances for paying transaction fees.
It makes implementation more generic.
Also i think, that basic assets module for tokens on generic-parachain is much more convenient now, because we do not need mint/burn and other complicated functionality, we had with generic-asset pallet, which is already removed from the frame.

I see you a few things here, let me try to understand what you mean. 
- assets module - we discussed this in a separate thread and I think we agreed to move to this for when we upgrade, I suggest then we do it for this version now. 
- For staking and paying for transaction fees, we will use balances

File: pallets/token-dealer/src/lib.rs [hunk 1]
```diff
@@ -1,155 +1,263 @@
...
-		}
-	}
+    fn handle_downward_message(msg: &DownwardMessage) {
+        match msg {
+            DownwardMessage::TransferInto(dest, relay_amount, remark) => {
+                let dest: T::AccountId = convert_hack(&dest);
+                let relay_amount: BalanceOf<T> = convert_hack(relay_amount);
+                let relay_account = RelayId::default().into_account();
+                // remark has a concerte type [u8; 32]
+                let asset_id: Option<AssetIdOf<T>> = convert_hack(remark);
```
Comment (iorveth) [inline pallets/token-dealer/src/lib.rs:195]: Are you sure, that it is possible to put asset_id in remark?

File: pallets/token-dealer/src/lib.rs [hunk 2]
```diff
@@ -1,155 +1,263 @@
...
+    fn handle_xcmp_message(
+        src: ParaId,
+        msg: &XCMPMessage<T::AccountId, BalanceOf<T>, AssetIdOf<T>>,
+    ) {
+        match msg {
+            XCMPMessage::TransferToken(dest, amount, asset_id) => {
+                let para_account = src.clone().into_account();
+
+                let res = match asset_id {
+                    Some(id) => <assets::Module<T>>::make_transfer(
```
Comment (iorveth) [inline pallets/token-dealer/src/lib.rs:236]: Shouldn`t we refactor it, as above?
```
                let para_account = T::Lookup::unlookup(para_account.clone());
<assets::Module<T>>::transfer(
		}	                    origin, id, para_account, convert_hack(&amount)

                )?;
```

**Review by iorveth — CHANGES_REQUESTED** [review]

File: pallets/token-dealer/src/lib.rs (hunk 1 above)
Comment (whalelephant) [reply pallets/token-dealer/src/lib.rs:195]: quite certain, my understanding is that Remark is just any 32btyes data, and as long as it is scale-encoded, which we use to decode it, it is fine. Like the RMK nft uses it too.

**Review by iorveth — APPROVED** [review]
LGTM

**Pull request merged**